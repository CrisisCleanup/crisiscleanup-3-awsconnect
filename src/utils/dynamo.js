/**
 * dynamo.js
 * DynamoDB Util
 */

import AWS from 'aws-sdk';
import AmazonDaxClient from 'amazon-dax-client';

let daxClient, ddbClient, dynamoClient;

export const dynamoOptions = () => {
  const isLocal = process.env.SLS_STAGE === 'local'; // serverless-offline
  let opts = {};
  if (isLocal) {
    opts = {
      region: 'localhost',
      endpoint:
        process.env.IS_OFFLINE === 'TUNNEL'
          ? 'http://marsdynamo.crisiscleanup.io'
          : 'http://localstack:4566',
    };
  }
  console.log('Dynamo Endpoints configured:', opts);
  console.log('DAX EP:', process.env.AWS_DAX_ENDPOINT);
  return opts;
};

export const TABLES = {
  AGENTS: {
    name: `connect-agents-table-${process.env.SLS_STAGE}`,
    hash: 'agent_id',
  },
  CONTACTS: {
    name: `connect-contacts-table-${process.env.SLS_STAGE}`,
  },
  CLIENTS: {
    name: `connect-clients-table-${process.env.SLS_STAGE}`,
  },
  METRICS: {
    name: `connect-metrics-table-${process.env.SLS_STAGE}`,
  },
};

export const DynamoTable = ({ name, bypassCache = false } = {}) => {
  const isLocal = process.env.SLS_STAGE === 'local'; // serverless-offline
  if (isLocal || bypassCache) {
    if (!dynamoClient) {
      dynamoClient = new AWS.DynamoDB({
        ...dynamoOptions(),
      });
    }
    return dynamoClient;
  }
  daxClient = new AmazonDaxClient({
    endpoints: [process.env.AWS_DAX_ENDPOINT],
  });
  return daxClient;
};

export const DynamoClient = ({ name } = {}) => {
  if (!daxClient) {
    DynamoTable();
  }
  if (dynamoClient) {
    if (!ddbClient) {
      ddbClient = new AWS.DynamoDB.DocumentClient({
        ...dynamoOptions(),
      });
    }
    return ddbClient;
  }
  if (!ddbClient) {
    ddbClient = new AWS.DynamoDB.DocumentClient({
      service: daxClient,
      ...dynamoOptions(),
    });
  }
  return ddbClient;
};

export const normalize = (record) => AWS.DynamoDB.Converter.unmarshall(record);

export const AttrExpression = ({ key = 's', name = 'state', value } = {}) => ({
  ExpressionAttributeNames: {
    [`#${key.toUpperCase()}`]: name,
  },
  ExpressionAttributeValues: {
    [`:${key.toLowerCase()}`]: value,
  },
});

export const Expressions = (exps) => {
  const finalExp = {
    ExpressionAttributeNames: {},
    ExpressionAttributeValues: {},
  };
  exps.forEach(({ valueOnly = false, keyOnly = false, ...exp } = {}) => {
    if (exp.value || typeof exp.value === 'number' || keyOnly || valueOnly) {
      const result = AttrExpression(exp);
      if (!valueOnly) {
        finalExp.ExpressionAttributeNames = {
          ...finalExp.ExpressionAttributeNames,
          ...result.ExpressionAttributeNames,
        };
      }
      if (!keyOnly) {
        finalExp.ExpressionAttributeValues = {
          ...finalExp.ExpressionAttributeValues,
          ...result.ExpressionAttributeValues,
        };
      }
    }
  });
  if (!Object.keys(finalExp.ExpressionAttributeNames).length) {
    return { ExpressionAttributeValues: finalExp.ExpressionAttributeValues };
  }
  if (!Object.keys(finalExp.ExpressionAttributeValues).length) {
    return { ExpressionAttributeNames: finalExp.ExpressionAttributeNames };
  }
  return finalExp;
};

// TTL Filter
export const expiredFilter = (
  { ExpressionAttributeNames = {}, ExpressionAttributeValues = {} } = {},
  { condition = false } = {},
) => {
  const newExp = {
    ...Expressions([
      { key: 'now', value: Math.floor(Date.now() / 1000), valueOnly: true },
      { key: 'x', name: 'ttl', keyOnly: true },
    ]),
  };
  return {
    ExpressionAttributeNames: {
      ...ExpressionAttributeNames,
      ...newExp.ExpressionAttributeNames,
    },
    ExpressionAttributeValues: {
      ...ExpressionAttributeValues,
      ...newExp.ExpressionAttributeValues,
    },
    [condition ? 'ConditionExpression' : 'FilterExpression']: '#X > :now',
  };
};
