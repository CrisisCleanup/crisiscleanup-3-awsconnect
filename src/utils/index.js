/**
 * index.js
 * Utils
 */

import axios from 'axios';

export const ENDPOINT = (isDev = null) => {
  const ep = {
    auth: `Token ${process.env.CC_AUTH_TOKEN}`,
    baseUrl: process.env.CC_API_BASE_URL,
  };
  // AWS connect likes to cast inputs as strings
  const dev = isDev || Boolean(Number(process.env.IS_DEV));
  if (dev) {
    ep.auth = `Token ${process.env.CC_DEV_AUTH_TOKEN}`;
    ep.baseUrl = process.env.CC_DEV_API_BASE_URL;
  }
  return ep;
};

export const configureEndpoint = (isDev = null) => {
  const apiConfig = ENDPOINT(isDev);
  const { baseUrl, auth } = apiConfig;
  axios.defaults.baseURL = baseUrl;
  axios.defaults.headers.common.Authorization = auth;
  return axios;
};
