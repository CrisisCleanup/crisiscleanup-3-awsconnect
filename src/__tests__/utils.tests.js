/**
 * utils.tests.js
 * Utils Tests
 */

// Environment based tests setup based on:
// https://stackoverflow.com/a/48042799
import { configureEndpoint } from '../utils';

describe('utils', () => {
  const OLD_ENV = process.env;

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...OLD_ENV };
  });

  afterEach(() => {
    process.env = OLD_ENV;
  });

  it('should set axios defaults correctly', () => {
    process.env.CC_AUTH_TOKEN = 'abc123';
    process.env.CC_API_BASE_URL = 'http://api.com';

    // with only base
    const ax = configureEndpoint();
    expect(ax.defaults).toMatchSnapshot();
  });
});
