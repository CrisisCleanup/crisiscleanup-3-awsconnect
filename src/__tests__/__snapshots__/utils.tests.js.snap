// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils should set axios defaults correctly 1`] = `
Object {
  "adapter": [Function],
  "baseURL": "http://api.io",
  "headers": Object {
    "common": Object {
      "Accept": "application/json, text/plain, */*",
      "Authorization": "Token abc123",
    },
    "delete": Object {},
    "get": Object {},
    "head": Object {},
    "patch": Object {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "post": Object {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "put": Object {
      "Content-Type": "application/x-www-form-urlencoded",
    },
  },
  "maxContentLength": -1,
  "timeout": 0,
  "transformRequest": Array [
    [Function],
  ],
  "transformResponse": Array [
    [Function],
  ],
  "validateStatus": [Function],
  "xsrfCookieName": "XSRF-TOKEN",
  "xsrfHeaderName": "X-XSRF-TOKEN",
}
`;
